### PR TITLE
Resolve warnings in consensus lib builds.

### PR DESCRIPTION
**Pull request by evoskuil**
These changes are limited to issues affecting the consensus library build. As a library intended for developer use it should build cleanly on all intended platforms. These changes should not have any effect on execution.

### CODE REVIEW

File: src/serialize.h [hunk 1]
```diff
@@ -187,20 +187,20 @@ void WriteCompactSize(Stream& os, uint64_t nSize)
...
-        unsigned short xSize = nSize;
+        unsigned short xSize = static_cast<unsigned short>(nSize);
         WRITEDATA(os, chSize);
         WRITEDATA(os, xSize);
     }
     else if (nSize <= std::numeric_limits<unsigned int>::max())
     {
         unsigned char chSize = 254;
-        unsigned int xSize = nSize;
+        unsigned int xSize = static_cast<unsigned short>(nSize);
```
Comment (theuni) [inline src/serialize.h]: whoops.

**Comment by evoskuil** [comment]
@theuni That is understandable, but there is only one line affected by this in the consensus build. Maybe that could be treated as an isolated case, since it's quite trivial.

**Comment by theuni** [comment]
@evoskuil if anything, as long as we're counting on that behavior, the change should be more like:
```
#ifdef NDEBUG
#error this is unsupported and very unwise.
#endif
```

File: src/pubkey.cpp [hunk 2]
```diff
@@ -68,8 +68,8 @@ bool CPubKey::Decompress() {
         return false;
 #ifdef USE_SECP256K1
     int clen = size();
-    int ret = secp256k1_ecdsa_pubkey_decompress((unsigned char*)begin(), &clen);
-    assert(ret);
+    if (secp256k1_ec_pubkey_decompress((unsigned char*)begin(), &clen) == 0)
+        return false;
```
Comment (evoskuil) [inline src/pubkey.cpp]: I missed this in my review before writing up my comments. This is a behavior change I meant to discuss before pushing. It doesn't factor in to the consensus call, and appears to be in error.  I would appreciate it if someone with more background would comment before I pull this out of the pull request.

Comment (theuni) [inline src/pubkey.cpp]: This code (everything in #ifdef USE_SECP256k1) is currently unused, and should just be removed until it's actually supported.

**Comment by theuni** [comment]
I suspect that the crashes here are due to some wonky var shadowing issues in serialization. Assuming that's the case, those will need to be addressed before cleaning up unused vars.

**Comment by evoskuil** [comment]
@theuni WRT the test fails, thanks for the hint, I wouldn't have guessed that.

**Comment by evoskuil** [comment]
@theuni In the interest of resolving the other issues, I took up your suggestion and removed the sections of unsupported code: `NDEBUG` and `USE_SECP256K1`.

**Comment by luke-jr** [comment]
Not sure the NDEBUG should be removed. At least it compiles (unlike libsecp256k1 validation).

**Comment by evoskuil** [comment]
Are there any objections to @luke-jr's suggestion that we retain the fix to the `NDEBUG` warning in the one instance within the consensus code?

**Comment by gmaxwell** [comment]
Verification with libsecp256k1 is not supported at this time. I'm very concerned that more or less intended to be cosmetic changes believed to not effect execution have (above) broken consensus several times. Please take a step back here.

**Comment by evoskuil** [comment]
@gmaxwell The discussion on the issues with secp256k1 above resulted in the conditional incorporation that library being *removed* in this pull request.

I'm not sure what you mean by asking me to take a step back. Should I withdraw my pull request because it wasn't initially perfect? I'd be happy to do so if someone else would kindly fix the warnings and breaks.

**Comment by gmaxwell** [comment]
I'm not directing a specific action there. My comment there was to point out that your changes were dangerous and you should probably reconsider your approach, and try to understand the root cause of the prior errors. Review cannot catch all errors, and if you continue to iterate against review and CI you will eventually get something that passes them but still may be broken.  You may want to consider breaking up your changes into smaller sets which are provably equivalent (e.g. produce identical object code).

With respect to NDEBUG. It is not a supported build configuration at this time. There is specific code in the codebase to actively prevent compilation with NDEBUG set:
'''
#if defined(NDEBUG)
# error "Bitcoin cannot be compiled without assertions."
#endif
'''

Assertions are not used for general debugging or error handling in the Bitcoin Core codebase. They are used to test critical invariants which are believed to always be true (on a non faulty compile, on not faulty hardware) where their undetected violation would likely result in a security vulnerability.  Historically they were sometimes used in ways in which the expression had side effects and would cause actual failure if removed. Although I believe I have removed all the uses side-effects the system has not been recently audited for them, and some could have been reintroduced (and I note, an overly aggressive avoidance of warnings is one of the things that can result in introducing side effects into assertion statements).

The commit that splits up a statement with NDEBUG ifdefs is not an approach we would be likely to accept, particularly not to silence an written but not read variable warning in an actively unsupported build configuration. On top of a general good practice avoidance of conditional compilation except where strictly needed for portability, performing a byte-saving line split is a messy approach which could lead to serious bugs in the future.

**Comment by evoskuil** [comment]
There was discussion on the NDBUG question above. I understand how it works. I've reverted it twice to try and satisfy maintainers. I realize it's a crappy fix to a crappy problem. We use a different approach in libbitcoin but I wanted to keep it simple/local here. I really don't care how or if it's resolved.

The consensus lib is presumably for developers. As I said at the outset, it should build cleanly (and not fail to build on major platforms). As a maintainer of one of the platforms that might use this build I've made a considerable effort to point out the issues. Given that these trivial changes are dangerous I will just suppress all warnings and move on.

**Comment by theuni** [comment]
@evoskuil Some of these are still useful, I think, if you're interested in attacking them in smaller chunks.

- The CSHA256 changes look good now that they're fixed.
- Fixed missing include is obviously fine if it's necessary
- The extra comma/semicolon fixes look good since iirc they're required for strict standard compliance.

Also personally I like the changes to make casts explicit, but I'm not sure if others would agree there.

I'd drop the unused params and NDEBUG changes though.

**Pull request closed without merging**